components/org.wso2.carbon.database.utils/src/main/java/org/wso2/carbon/database/utils/jdbc/JdbcTemplate.java

```diff
@@ -0,0 +1,374 @@
+import org.wso2.carbon.database.utils.jdbc.exceptions.DataAccessException;
+import org.wso2.carbon.database.utils.jdbc.exceptions.TransactionException;
+
+import java.sql.Connection;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.util.ArrayList;
+import java.util.List;
+import javax.sql.DataSource;
+
+/**
+ * A utility class to call JDBC with lambda expressions. This class simplifies the usage of JDBC and helps to avoid
+ * common errors.
+ * Using this class we can execute insert, select, update and delete queries by providing necessary parameters to the
+ * methods. Transaction support is also available with this implementation for the query executions. All the SQL
+ * operations performed by this class are logged at the debug level.
+ */
+public class JdbcTemplate {
+
+    private static final Logger log = LoggerFactory.getLogger(JdbcTemplate.class);
+    private static final String ID = "ID";
+    private static String driverName;
+    private DataSource dataSource;
+
+    public JdbcTemplate(DataSource dataSource) {
+
+        this.dataSource = dataSource;
+    }
+
+    public <T extends Object, E extends Exception> T withTransaction(ExecuteCallable<T> callable) throws
+            TransactionException, E {
+
+        TransactionEntry transactionEntry = TransactionManager.getTransactionEntry();
+        if (transactionEntry.getTransactionDepth() == 0) {
+
+            try {
+                Connection connection = dataSource.getConnection();
+                transactionEntry.setAutoCommitState(connection.getAutoCommit());
+                connection.setAutoCommit(false);
+                transactionEntry.setConnection(connection);
+            } catch (SQLException e) {
+                throw new TransactionException(JdbcConstants.ErrorCodes.ERROR_CODE_GETTING_CONNECTION_FAILURE.
+                        getErrorMessage(), e);
+            }
+        }
+        transactionEntry.incrementTransactionDepth();
+        TransactionManager.setTransactionEntry(transactionEntry);
+        try {
+            T result = callable.get(new Template<>());
+            transactionEntry.decrementTransactionDepth();
+            if (transactionEntry.getTransactionDepth() == 0) {
+                transactionEntry.getConnection().commit();
+            }
+            return result;
+        } catch (Throwable t) {
+            //We catch any exception and rollback the transaction
+            if (log.isDebugEnabled()) {
+                log.debug("Could not commit the transaction.", t);
+            }
+            transactionEntry.decrementTransactionDepth();
+            if (transactionEntry.getTransactionDepth() == 0) {
+                try {
+                    transactionEntry.getConnection().rollback();
+                } catch (SQLException e) {
+                    if (log.isDebugEnabled()) {
+                        log.debug("Could not rollback the transaction. - " + e.getMessage(), e);
```

Comment:
log.debug("message", e) should print the message in "e". There is no need to manually append the e.getMessage().

Reply:
Done.